gamestonk_terminal/cryptocurrency/onchain/bitquery_model.py

```diff
@@ -0,0 +1,638 @@
+    pd.DataFrame
+        Token volume on Decentralized Exchanges
+    """
+
+    if trade_amount_currency not in CURRENCIES:
+        trade_amount_currency = "USD"
+
+    token_address = find_token_address(token)
+    query = f"""
+        {{
+           ethereum {{
+            dexTrades(
+              baseCurrency: {{is:"{token_address}"}}
+            ) {{
+                  baseCurrency{{
+        symbol
+      }}
+              exchange {{
+              name
+              fullName
+              }}
+              count
+              tradeAmount(in: {trade_amount_currency})
+
+            }}
+            }}
+            }}
+
+        """
+
+    data = query_graph(BQ_URL, query)
+    if not data:
+        return pd.DataFrame()
+
+    df = pd.json_normalize(data["ethereum"]["dexTrades"])[
+        ["exchange.fullName", "baseCurrency.symbol", "tradeAmount", "count"]
+    ]
+    df.columns = ["exchange", "coin", "tradeAmount", "trades"]
+    return df[~df["exchange"].str.startswith("<")].sort_values(
+        by="tradeAmount", ascending=False
+    )
+
+
+def get_ethereum_unique_senders(interval: str = "day", limit: int = 90) -> pd.DataFrame:
+    """Get number of unique ethereum addresses which made a transaction in given time interval.
+
+    Parameters
+    ----------
+    interval: str
+        Time interval in which count unique ethereum addresses which made transaction. day, month or week.
+    limit: int
+        Number of records for data query.
+
+    Returns
+    -------
+    pd.DataFrame
+        Unique ethereum addresses which made a transaction
+    """
+
+    intervals = {
+        "day": 1,
+        "month": 30,
+        "week": 7,
+    }
+
+    if interval not in intervals.keys():
+        interval = "day"
+
+    days = min(limit * intervals[interval], 90)
+
+    dt = (datetime.date.today() - datetime.timedelta(days)).strftime("%Y-%m-%d")
+
+    query = f"""
+     {{
+          ethereum(network: ethereum) {{
+            transactions(options: {{desc: "date.date"}}, date: {{since: "{dt}"}}) {{
+              uniqueSenders: count(uniq: senders)
+              date {{
+                date:startOfInterval(unit: {interval})
+              }}
+                avgGasPrice: gasPrice(calculate: average)
+                  medGasPrice: gasPrice(calculate: median)
+                  maxGasPrice: gasPrice(calculate: maximum)
+                  transactions: count
+            }}
+          }}
+        }}
+        """
+
+    data = query_graph(BQ_URL, query)
+    if not data:
+        return pd.DataFrame()
+
+    df = pd.DataFrame(data["ethereum"]["transactions"])
+    df["date"] = df["date"].apply(lambda x: x["date"])
+    return df[
+        [
+            "date",
+            "uniqueSenders",
+            "transactions",
+            "avgGasPrice",
+            "medGasPrice",
+            "maxGasPrice",
+        ]
+    ]
+
+
+def get_most_traded_pairs(
+    network: str = "ethereum", exchange: str = "Uniswap", limit: int = 90
```

Comment:
So, for the query to bitquery api case sensitive exchange name is important, but I will try to make it possible that user don't need to take care about it.

Reply:
Have a look at how Countries are selected in the `sector industry analysis` menu (it's still WIP but it looks and feels very nice). In the `call_country` function there is a code block that adds autocomplete for the country selection and substitutes all lowercase string with correct capitalized string.
The code is currently in a branch, but if it changes 'cause the PR is merged you'll find in it main
https://github.com/GamestonkTerminal/GamestonkTerminal/blob/d4babe3dece5fcbd71f31fc0e1addc2ab1184869/gamestonk_terminal/stocks/sector_industry_analysis/sia_controller.py#L536

Reply:
Thanks for your comment, I implemented some stuffs with difflib in crypto menu, so I have some basic knowledge how to implement it. I took a look on `call_country` function, but I don't know if I understand correctly how this autocompleter works. Later I will try to check branch with this feature on my local computer, to see how it works. But in the meantime, if you have time you can take a look on my solution. :)

I think a little issue there is that we explicitly add list of possible exchanges into argparser choice parameter. So if we want to use difflib to get best match for what user wrote, we need to resign from choice parameter which is an array of possible exchanges. So I don't know what it will be a better experience for user: to have difflib in the background, but not see possible exchanges with -h parameter, or to have listed exchanges with lowercase, and in the background to have a mapping, that will translate lower case string into what api expects. For now I provide b) solution with lowercase and background mapping.

Reply:
@DidierRLopes can you have a look here if it's possible to add autocomplete for exchanges the same way the country selection is done in sia?

Reply:
Yup, so this is how you want to do it:
<img width="752" alt="Screenshot 2021-12-06 at 18 32 06" src="https://user-images.githubusercontent.com/25267873/144902105-dc6c6985-e29c-42ca-8924-0a3794d74fe1.png">

In your case the choices should be 
```
choices["mt"]= {c: None for c in possible_exchanges}
```

Note that this will only happen when you do `mt unis...` for instance but it's still very helpful.


Also, look into how I implemented this:
```
 def call_country(self, other_args: List[str]):
        """Process country command"""
        parser = argparse.ArgumentParser(
            add_help=False,
            formatter_class=argparse.ArgumentDefaultsHelpFormatter,
            prog="country",
            description="See existing countries, or set country if arg specified",
        )
        parser.add_argument(
            "-n",
            "--name",
            type=str,
            dest="name",
            nargs="+",
            help="country to select",
        )
        if other_args:
            if "-" not in other_args[0]:
                other_args.insert(0, "-n")
        ns_parser = parse_known_args_and_warn(parser, other_args)
         if not ns_parser:
             return

         possible_countries = financedatabase_model.get_countries(
             industry=self.industry, sector=self.sector
         )

         if ns_parser.name:
             if " ".join(ns_parser.name) in possible_countries:
                 self.country = " ".join(ns_parser.name)
             else:
                 print(f"Country '{' '.join(ns_parser.name)}' does not exist.")

                 similar_cmd = difflib.get_close_matches(
                     " ".join(ns_parser.name),
                     possible_countries,
                     n=1,
                     cutoff=0.75,
                 )
                if similar_cmd:
                    print(f"Replacing by '{similar_cmd[0]}'")
                    self.country = similar_cmd[0]
                 else:
                     similar_cmd = difflib.get_close_matches(
                         " ".join(ns_parser.name),
                         possible_countries,
                         n=1,
                         cutoff=0.5,
                     )
                     if similar_cmd:
                         print(f"Did you mean '{similar_cmd[0]}'?")

         else:
             for country in possible_countries:
                 print(country)
```

You are right when you say that the `choices=...` allows to retrieve an error straight if user doesn't provide correct argument. BUT with this implementation you test difflib similarity, if it is above 0.75 you go ahead and replace, If it is between 0.5 and 0.75 you ask "did you mean XXXX" and if it's below you can say something like "XXX not found. Use one of the following: .....". This way you are not as harsh saying that it doesn't exist even if the difference is just a capital letter or so. 

Does that make sense?

Reply:
Didier, thank you for comment. Will try to implement it for decentralized exchanges.